internals-js/src/extractSubgraphsFromSupergraph.ts

```diff
@@ -625,51 +577,360 @@ function errorToString(e: any,): string {
+    // to some of the types.
+    for (const type of subgraph.schema.types()) {
+      switch (type.kind) {
+        case 'ObjectType':
+        case 'InterfaceType':
+        case 'InputObjectType':
+          if (!type.hasFields()) {
+            // Note that we have to use removeRecursive or this could leave the subgraph invalid. But if the
+            // type was not in this subgraphs, nothing that depends on it should be either.
+            type.removeRecursive();
+          }
+          break;
+        case 'UnionType':
+          if (type.membersCount() === 0) {
+            type.removeRecursive();
+          }
+          break;
+      }
+    }
+
+    // Lastly, we add all the "executable" directives from the supergraph to each subgraphs, as those may be part
+    // of a query and end up in any subgraph fetches. We do this "last" to make sure that if one of the directive
+    // use a type for an argument, that argument exists.
+    // Note that we don't bother with non-executable directives at the moment since we've don't extract their
+    // applications. It might become something we need later, but we don't so far.
+    for (const definition of allExecutableDirectives) {
+      // Note that we skip any potentially applied directives in the argument of the copied definition, because as said
+      // in the comment above, we haven't copied type-system directives. And so far, we really don't care about those
+      // applications.
+      copyDirectiveDefinitionToSchema({
+        definition,
+        schema: subgraph.schema,
+        copyDirectiveApplicationsInArguments: false,
+        locationFilter: (loc) => isExecutableDirectiveLocation(loc),
+      });
+    }
+  }
+
+  // TODO: Not sure that code is needed anymore (any field necessary to validate an interface will have been marked
+  // external)?
+
+  // We now make a pass on every field of every interface and check that all implementers do have that field (even if
+  // external). If not (which can happen because, again, the v0.1 spec had no information on where an interface was
+  // truly defined, so we've so far added them everywhere with all their fields, but some fields may have been part
+  // of an extension and be only in a few subgraphs), we remove the field or the subgraph would be invalid.
+  for (const subgraph of subgraphs) {
+    for (const itf of subgraph.schema.interfaceTypes()) {
+      // We only look at objects because interfaces are handled by this own loop in practice.
+      const implementations = itf.possibleRuntimeTypes();
+      for (const field of itf.fields()) {
+        if (!implementations.every(implem => implem.field(field.name))) {
+          field.remove();
+        }
+      }
+      // And it may be that the interface wasn't part of the subgraph at all!
+      if (!itf.hasFields()) {
+        itf.remove();
+      }
+    }
+  }
```

Comment:
Maybe doesn't have to be this PR but should we remove this as suggested by the comment? Looks like tests pass without it but that's the extent of my understanding of its importance. Another potentially hundreds/thousands of iterations.

Reply:
Good call. When I noticed this, I was doing something unrelated and didn't wanted to risk breaking something, hence the TODO instead of removing it right away. But this patch is a refactor, so it's as good a place as any to make the removal. And double checked that this code indeed is not needed anymore (as the TODO says, the code that adds missing external fields that is called before that will ensure there is no implementation missing a field needed for an interface).

But I'll also point out that I rebased this/will merge this against `next` because this is a refactor, not a bug fix, so really belong in `next` (and the other PR that rely on this also really belong in `next` anyway). Which is all the more reason not to be shy about removing this.